### PR TITLE
HDDS-5390. reconPipelineReportHandler should not retry when pipeline not found

### PR DESCRIPTION
**Pull request by JacksonYao287**
## What changes were proposed in this pull request?

if recon receives a pipelineReport , the reconPipelineReportHandler will confirm it with scm. if scm does not find this pipeline , it will throw a pipelineNotFoundExecption. today, if this happen, reconPipelineReportHandler will try again and again. this behavior is not expected. the right behavior is that if this happen, reconPipelineReportHandler should just ignore this report and go ahead. 

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-5390

## How was this patch tested?

no need no test

### CODE REVIEW

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelineReportHandler.java
Comment (bharatviswa504) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelineReportHandler.java:96]: Does nt recon pipelineHandler processPipelineReport handled this already?


ReconPipelineReportHandler.java
```
    Pipeline pipeline = null;
    try {
      pipeline = reconPipelineManager.getPipeline(pipelineID);
    } catch (PipelineNotFoundException ex) {
      LOG.warn("Pipeline {} not found in Recon.", pipelineID);
      return;
    }
````

File: hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/client/HddsClientUtils.java
Comment (bharatviswa504) [inline hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/client/HddsClientUtils.java:74]: And also I see this list is being used in KeyOutputStream and XceiverClientRatis.
Not understood why we need this to be added here.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelineReportHandler.java
Comment (JacksonYao287) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelineReportHandler.java:96]: what i want to do here is to handle this 
```
      Pipeline pipelineFromScm =
          scmServiceProvider.getPipeline(report.getPipelineID());
```
here , a scmclient is used to confirm the pipeline existence to scm, and this will take a grpc call.

File: hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/client/HddsClientUtils.java
Comment (JacksonYao287) [reply hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/client/HddsClientUtils.java:74]: when a scmclient  is taking a grpc call, if any exception in this list thrown,  it will stop retry, or else it will try for a maxcount at a specified interval , which are set at `SCMClientConfig.java`. here , i want `scmclient.getpipeline()` not to retry if PipelineNotFoundException is threw.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelineReportHandler.java
Comment (bharatviswa504) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelineReportHandler.java]: Do you mean hadoop rpc which uses failover proxy provider i believe

File: hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/client/HddsClientUtils.java
Comment (bharatviswa504) [reply hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/client/HddsClientUtils.java:74]: And also we don't use this, we use the list specified in 
```
  private static final List<Class<? extends Exception>>
      NON_RETRIABLE_EXCEPTION_LIST =
      ImmutableList.<Class<? extends Exception>>builder()
          .add(SCMException.class)
          .add(NonRetriableException.class)
          .build();
```
SCMHAUtils.java

And also you can refer SCMHAUtils#getRetryAction for this.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconPipelineReportHandler.java
Comment (bharatviswa504) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconPipelineReportHandler.java:71]: From SCMClient, I don't think we get PipelineNotFoundException I believe we get exception as IOException.

```

  public static IOException getRemoteException(ServiceException se) {
    Throwable e = se.getCause();
    if (e == null) {
      return new IOException(se);
    }
    return e instanceof IOException ? (IOException) e : new IOException(se);
  }
```
```

  private ScmContainerLocationResponse submitRequest(
      StorageContainerLocationProtocolProtos.Type type,
      Consumer<Builder> builderConsumer) throws IOException {
    final ScmContainerLocationResponse response;
    try {
      Builder builder = ScmContainerLocationRequest.newBuilder()
          .setCmdType(type)
          .setVersion(CURRENT_VERSION)
          .setTraceID(TracingUtil.exportCurrentSpan());
      builderConsumer.accept(builder);
      ScmContainerLocationRequest wrapper = builder.build();

      response = submitRpcRequest(wrapper);
    } catch (ServiceException ex) {
      throw ProtobufHelper.getRemoteException(ex);
    }
    return response;
  }
```

Can you add a simple test with a cluster start, and see do we get PipelineNotFoundException?

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelineReportHandler.java
Comment (JacksonYao287) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelineReportHandler.java]: yea, it will use , please take a look at [this](https://issues.apache.org/jira/secure/attachment/13027279/%E6%88%AA%E5%B1%8F2021-06-25%20%E4%B8%8B%E5%8D%884.18.35.png)

File: hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/client/HddsClientUtils.java
Comment (JacksonYao287) [reply hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/client/HddsClientUtils.java:74]: do you mean if i want to forbid this retry here for pipelineNotFoundException , i should set it here?

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconPipelineReportHandler.java
Comment (JacksonYao287) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconPipelineReportHandler.java:71]: please see the log i mention [above ](https://github.com/apache/ozone/pull/2371#discussion_r658695134), the server side throw a PipelineNotFoundException, so if i do not misunderstand, the client side will also get a PipelineNotFoundException. please correct me if i am wrong

Comment (bharatviswa504) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconPipelineReportHandler.java:71]:
````
    try {
      cluster.getStorageContainerLocationClient().getPipeline(HddsProtos.PipelineID.newBuilder().setId(UUID.randomUUID().toString()).build());
    }  catch (IOException ex) {
      Assert.assertTrue(ex instanceof RemoteException);
      PipelineNotFoundException exception = (PipelineNotFoundException) ((RemoteException) ex).unwrapRemoteException(PipelineNotFoundException.class);
      Assert.assertTrue(exception instanceof PipelineNotFoundException);
    }
````

We donot get PipelineNotFoundException, we get RemoteException wrapped with PipelineNotFoundException. We need to unwrap and check if it is PipelineNotFoundException.

**Review by bharatviswa504 — CHANGES_REQUESTED** [review]
Need to fix Exception handling

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconPipelineReportHandler.java
Comment (JacksonYao287) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconPipelineReportHandler.java:71]: thanks for the review ,  i will fix this

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconPipelineReportHandler.java
Comment (bharatviswa504) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconPipelineReportHandler.java:76]: Not understood why we need this assert here. It can also be any other exception right? (Lets say when SCM's are down)

Comment (JacksonYao287) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconPipelineReportHandler.java:76]: yea, thanks , i made a mistake here, i will change this

Comment (JacksonYao287) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconPipelineReportHandler.java]: @bharatviswa504  PTAL, thanks!

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconPipelineReportHandler.java
Comment (bharatviswa504) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconPipelineReportHandler.java:76]: In this case do we need to throw PipelineNotFoundException??

Comment (bharatviswa504) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconPipelineReportHandler.java:76]: And also avoid loggig??

catch(NotLeaderException ex) {
        // Avoid NotLeaderException logging which happens when processing
        // pipeline report on followers.
      } catch(PipelineNotFoundException e) {
      // When pipeline not found donot log??
      }catch(IOException e) {
        LOGGER.error("Could not process pipeline report={} from dn={}.",
            report, dn, e);
      }

**Review by bharatviswa504 — COMMENTED** [review]
I have one question

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconPipelineReportHandler.java
Comment (JacksonYao287) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconPipelineReportHandler.java:76]: thank for the review , i have updated the PR , please take a look

**Review by bharatviswa504 — APPROVED** [review]
+1 LGTM
Thank You @JacksonYao287 for the fix

**Comment by JacksonYao287** [comment]
thanks @bharatviswa504  for the careful review, i merge the latest master branch to trigger CI again and wait for a clean CI before merge

**Comment by bharatviswa504** [comment]
Thank You @JacksonYao287 for the fix.

**Pull request merged**
